crates/polars-plan/src/plans/ir/mod.rs

```diff
@@ -203,6 +210,844 @@ impl IRPlan {
+            },
+            IR::HConcat {
+                inputs,
+                schema,
+                options,
+            } => {
+                let new_inputs: Vec<_> = inputs
+                    .iter()
+                    .map(|&input| {
+                        Self::convert_to_placeholder(input, old_arena, new_arena, placeholder_id)
+                    })
+                    .collect();
+                IR::HConcat {
+                    inputs: new_inputs,
+                    schema: schema.clone(),
+                    options: *options,
+                }
+            },
+            IR::ExtContext {
+                input,
+                contexts,
+                schema,
+            } => {
+                let new_input =
+                    Self::convert_to_placeholder(*input, old_arena, new_arena, placeholder_id);
+                let new_contexts: Vec<_> = contexts
+                    .iter()
+                    .map(|&ctx| {
+                        Self::convert_to_placeholder(ctx, old_arena, new_arena, placeholder_id)
+                    })
+                    .collect();
+                IR::ExtContext {
+                    input: new_input,
+                    contexts: new_contexts,
+                    schema: schema.clone(),
+                }
+            },
+            IR::Sink { input, payload } => {
+                let new_input =
+                    Self::convert_to_placeholder(*input, old_arena, new_arena, placeholder_id);
+                IR::Sink {
+                    input: new_input,
+                    payload: payload.clone(),
+                }
+            },
+            IR::SinkMultiple { inputs } => {
+                let new_inputs: Vec<_> = inputs
+                    .iter()
+                    .map(|&input| {
+                        Self::convert_to_placeholder(input, old_arena, new_arena, placeholder_id)
+                    })
+                    .collect();
+                IR::SinkMultiple { inputs: new_inputs }
+            },
+            #[cfg(feature = "merge_sorted")]
+            IR::MergeSorted {
+                input_left,
+                input_right,
+                key,
+            } => {
+                let new_left =
+                    Self::convert_to_placeholder(*input_left, old_arena, new_arena, placeholder_id);
+                let new_right = Self::convert_to_placeholder(
+                    *input_right,
+                    old_arena,
+                    new_arena,
+                    placeholder_id,
+                );
+                IR::MergeSorted {
+                    input_left: new_left,
+                    input_right: new_right,
+                    key: key.clone(),
+                }
+            },
+            // Nodes without inputs - just clone them
+            IR::PlaceholderScan { .. } => ir.clone(),
+            IR::Invalid => ir.clone(),
+        };
+        new_arena.add(new_ir)
+    }
+
+    pub fn bind_data(
```

Comment:
Do these methods (with `bind_df`) need to be public? If it's the case we probably want a docstring